### PR TITLE
Make long topic expandable.

### LINKED ISSUE #1957: Very long topic cannot be fully visible
We should limit the number of lines of the topic in the Toolbar and make it clickable to be fully displayed in a dedicated screen

So:
- [x] Remove the topic from the extended toolbar
- [x] Add a topic section in the room settings (above security section), with expand/collapse action, at top end of the "topic" header

See what has been done on iOS for reference (even if it is not strictly similar): https://github.com/vector-im/element-ios/pull/3724

### PR DESCRIPTION
**Pull request by onurays**
Fixes #1957 

WhatsApp Image 2020-10-09 at 14 48 44

WhatsApp Image 2020-10-09 at 14 49 00

### CODE REVIEW

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+                holder.view.setOnClickListener {
+                    if (isExpanded) {
+                        collapse(holder)
+                    } else {
+                        expand(holder)
+                    }
+                }
+            } else {
+                holder.arrow.isVisible = false
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:60]: if this Item would be recycled, the arrow could be Gone. You should make it visible in the `if` block

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            }
+        }
+    }
+
+    private fun expand(holder: Holder) {
+        ObjectAnimator.ofInt(holder.content, "maxLines", expandedLines)
+                .apply { duration = 200 }
+                .also { it.start() }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:68]: Simplify to 
```
.setDuration(200)
.start()
```

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        isExpanded = true
+    }
+
+    private fun collapse(holder: Holder) {
+        ObjectAnimator.ofInt(holder.content, "maxLines", maxLines)
+                .apply { duration = 200 }
+                .also { it.start() }
+
+        holder.content.ellipsize = TextUtils.TruncateAt.END
+                holder.arrow.setImageResource(R.drawable.ic_expand_more)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:81]: Format?

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+@EpoxyModelClass(layout = R.layout.item_expandable_textview)
+abstract class ExpandableTextItem : VectorEpoxyModel<ExpandableTextItem.Holder>() {
+
+    @EpoxyAttribute
+    lateinit var content: String
+
+    @EpoxyAttribute
+    var maxLines: Int = 3
+
+    private var isExpanded = false
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:39]: This should be kept in the State, but maybe it's not a big deal

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        }
+    }
+
+    private fun expand(holder: Holder) {
+        ObjectAnimator.ofInt(holder.content, "maxLines", expandedLines)
+                .apply { duration = 200 }
+                .also { it.start() }
+
+        holder.content.ellipsize = null
+        holder.arrow.setImageResource(R.drawable.ic_expand_less)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:71]: Please also update contentDescription. You can use

    <string name="merged_events_expand">expand</string>
    <string name="merged_events_collapse">collapse</string>

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        isExpanded = true
+    }
+
+    private fun collapse(holder: Holder) {
+        ObjectAnimator.ofInt(holder.content, "maxLines", maxLines)
+                .apply { duration = 200 }
+                .also { it.start() }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:78]: same remark, please simplify

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt [hunk 7]
```diff
@@ -57,6 +58,20 @@ class RoomProfileController @Inject constructor(
...
+
+        // Topic
+        roomSummary
+                .topic
+                .takeIf { it.isNotBlank() }
+                ?.let {
+                    buildProfileSection(stringProvider.getString(R.string.room_profile_section_topic))
+                    expandableTextItem {
+                        id("topic")
+                        content(roomSummary.topic)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt:70]: Replace by `content(it)`

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt [hunk 8]
```diff
@@ -57,6 +58,20 @@ class RoomProfileController @Inject constructor(
             return
         }
         val roomSummary = data.roomSummary() ?: return
+
+        // Topic
+        roomSummary
+                .topic
+                .takeIf { it.isNotBlank() }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt:65]: I would have used `isNotEmpty()`

File: vector/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -2124,6 +2124,7 @@
     <string name="direct_room_profile_not_encrypted_subtitle">Messages here are not end-to-end encrypted.</string>
     <string name="room_profile_encrypted_subtitle">Messages in this room are end-to-end encrypted.\n\nYour messages are secured with locks and only you and the recipient have the unique keys to unlock them.</string>
     <string name="direct_room_profile_encrypted_subtitle">Messages here are end-to-end encrypted.\n\nYour messages are secured with locks and only you and the recipient have the unique keys to unlock them.</string>
+    <string name="room_profile_section_topic">Topic</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:2127]: Please remove and use `room_settings_topic` which already exist and is already translated in many languages.

**Review by bmarty — CHANGES_REQUESTED** [review]
Thanks. A few remarks

File: vector/src/main/res/layout/view_stub_room_profile_header.xml [hunk 10]
```diff
@@ -54,27 +54,9 @@
...
         app:layout_constraintTop_toBottomOf="@+id/roomProfileNameView"
         tools:text="@sample/matrix.json/data/roomAlias" />
 
-    <TextView
-        android:id="@+id/roomProfileTopicView"
-        android:layout_width="0dp"
-        android:layout_height="wrap_content"
-        android:layout_marginStart="40dp"
-        android:layout_marginEnd="40dp"
-        android:autoLink="web"
```
Comment (bmarty) [inline vector/src/main/res/layout/view_stub_room_profile_header.xml:69]: We have lost this, please restore in the new layout

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt (hunk 4 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:39]: I am aware of it but I also don't think it is a big deal.

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt (hunk 1 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:60]: Done.

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt (hunk 2 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:68]: Done.

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt (hunk 3 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:81]: Done.

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt (hunk 5 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:71]: Done.

File: vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt (hunk 6 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/core/epoxy/ExpandableTextItem.kt:78]: Done.

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt (hunk 7 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt:70]: Done.

File: vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt (hunk 8 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/roomprofile/RoomProfileController.kt:65]: Done.

File: vector/src/main/res/values/strings.xml (hunk 9 above)
Comment (onurays) [reply vector/src/main/res/values/strings.xml:2127]: Done.

File: vector/src/main/res/layout/view_stub_room_profile_header.xml (hunk 10 above)
Comment (onurays) [reply vector/src/main/res/layout/view_stub_room_profile_header.xml:69]: Done.

**Comment by bmarty** [comment]
Thanks for the update!

File: vector/src/main/res/layout/item_expandable_textview.xml [hunk 11]
```diff
@@ -23,7 +23,12 @@
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
         android:layout_marginTop="4dp"
+        android:autoLink="web"
+        android:fontFamily="sans-serif"
+        android:gravity="center"
```
Comment (bmarty) [inline vector/src/main/res/layout/item_expandable_textview.xml:28]: I was just thinking of `autoLink`. I will revert Gravity later on.

File: vector/src/main/res/layout/view_stub_room_profile_header.xml (hunk 10 above)
Comment (bmarty) [reply vector/src/main/res/layout/view_stub_room_profile_header.xml:69]: You added back the attributes to the ImageView and not the TextView... I will fix it.

**Pull request merged**